### PR TITLE
[CUDA] Fix missing bounds check in `Softmax.cu`

### PR DESCRIPTION
**Pull request by eqy**
Uncovered by @ngimel, same as change in #144009 

cc @ptrblck @msaroufim @jerryzh168

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

**Review by cyyever — APPROVED** [review]

**Review by malfet — APPROVED** [review]
Would be nice to have test (or at least a test plan)

File: test/test_nn.py [hunk 1]
```diff
@@ -10498,6 +10498,17 @@ def test_softmax_double(self, device, dtype):
         expected_ones = F.log_softmax(logits, dim=1).exp().sum(dim=1)
         self.assertEqual(expected_ones, torch.ones_like(expected_ones))
 
+        # backward
+        logits = torch.randn(5, 513, dtype=dtype, device=device, requires_grad=True)
```
Comment (ngimel) [inline test/test_nn.py:10502]: does it actually go through that kernel?

Comment (eqy) [reply test/test_nn.py:10502]: it actually goes through the kernel but there's a heuristic determining the block size rather than hardcoding it to 512 and I couldn't find a case that actually causes size to go negative
